### ISSUE TITLE
The expect override is confused by tests.

### ISSUE THREAD

**Issue by IgorBelyi** [rust-lang/rust-clippy#16488]
### Summary

I'd like to start using `clippy::allow_attributes` but I'm running into a problem when some code is used in tests only. 

### Lint Name

unused

### Reproducer

I have the following code reproducing the issue.
```rust
#![warn(clippy::allow_attributes)]

#[expect(unused, reason = "Still testing")]
struct Test {}

fn main() {}

#[test]
fn test_test() {
    let _ = Test {};
}
```
It results in an unexpected warning:
```
-> src/bin/test_allow.rs:3:10
  |
3 | #[expect(unused, reason = "Still testing")]
  |          ^^^^^^
  |
  = note: Still testing
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```
If I comment out `expect` override the error changes to 
```
 --> src/bin/test_allow.rs:4:8
  |
4 | struct Test {}
  |        ^^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default
```
If I change `expect` to `allow` the error becomes
```
 --> src/bin/test_allow.rs:3:3
  |
3 | #[allow(unused, reason = "Still testing")]
  |   ^^^^^ help: replace it with: `expect`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#allow_attributes
note: the lint level is defined here
 --> src/bin/test_allow.rs:1:9
  |
1 | #![warn(clippy::allow_attributes)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^
```
The only way I found to avoid warning is to remove `#![warn(clippy::allow_attributes)]` and use `#[allow(unused)]` to suppress `unused` warning.

### Version

```text
rustc 1.93.0 (254b59607 2026-01-19)
binary: rustc
commit-hash: 254b59607d4417e9dffbc307138ae5c86280fe4c
commit-date: 2026-01-19
host: x86_64-apple-darwin
release: 1.93.0
LLVM version: 21.1.8
```

**Comment by zacknewman** [comment]
This isn't a Clippy issue or even `expect` issue. This is just a consequence of how attributes work. If you want to conditionally apply an attribute, you have no choice but to use [`cfg_attr`](https://doc.rust-lang.org/reference/conditional-compilation.html#the-cfg_attr-attribute). Something like below will accomplish what you want:

```rust
// These lint attributes apply unconditionally (e.g., they apply when the test
// harness is compiled).
#![deny(unfulfilled_lint_expectations, unused, reason = "example")]
// We conditionally `expect` `dead_code` (which is part of the `unused`
// group) when the test harness is not compiled.
#[cfg_attr(not(test), expect(dead_code, reason = "example"))]
struct Foo;
fn main() {}
#[test]
fn foo() {
    _ = Foo;
}
```

If you were to comment out `_ = Foo;` in `foo`, `cargo check --tests`/`cargo check --all-targets` would fail due to `dead_code`. Similarly if you were to use `Foo` in `main`, `cargo check --bins`/`cargo check --all-targets` would fail due to `unfulfilled_lint_expectations`.

Alternatively you can do something like below:

```rust
// Globally `allow` these lints when the test harness
// is compiled.
#![cfg_attr(
    test,
    allow(
        unfulfilled_lint_expectations,
        clippy::allow_attributes,
        reason = "example"
    )
)]
// Globally `deny` these lints when the test harness
// is not compiled.
#![cfg_attr(
    not(test),
    deny(
        unfulfilled_lint_expectations,
        clippy::allow_attributes,
        reason = "example"
    )
)]
// Globally `deny` these lints.
#![deny(unused, reason = "example")]
#[expect(dead_code, reason = "example")]
struct Foo;
fn main() {}
#[test]
fn foo() {
    _ = Foo;
}
```

FYI, it's more idiomatic to have your tests in a separate module. You can use [`clippy::tests_outside_test_module`](https://rust-lang.github.io/rust-clippy/master/index.html#tests_outside_test_module) to check for this if that's something you want.

**Comment by IgorBelyi** [comment]
@zacknewman Thank you for the detailed explanation. Combining with `cfg_attr(not(test),...)` is what I was missing. And thank you for the `clippy::tests_outside_test_module` suggestion I'll add it to our collection.
